### ISSUE TITLE
Possible bug in the design of `tf.keras.layers.Conv`

### ISSUE THREAD

**Issue by fabio12345** [tensorflow/tensorflow#23101]
**System information**
- TensorFlow version (you are using): 1.10
- Are you willing to contribute it (Yes/No): No (Don't know tf's code base well enough.)

**Describe the feature and the current behavior/state.**
Current behaviour of `tf.layers.Conv2D` is that
```
import tensorflow as tf

example = tf.zeros([1, 32, 32, 1])
example2 = tf.zeros([1, 41, 41, 1])

# Using a dilated convolution layer
convolution_op = tf.layers.Conv2D(4, 3, padding="SAME", dilation_rate=4)

print(convolution_op( example ))
print(convolution_op( example2 ))
```
causes the following exception
```
ValueError: Dimension size must be evenly divisible by 4 but is 49 for 'conv2d_2/SpaceToBatchND_1' (op: 'SpaceToBatchND') with input shapes: [1,41,41,1], [2], [2,2] and with computed input tensors: input[1] = <4 4>, input[2] = <[4 4][4 4]>.
```
The reason for this is that in the `build` function for `tf.keras.layers.Conv` (https://github.com/tensorflow/tensorflow/blob/r1.11/tensorflow/python/keras/layers/convolutional.py) the convolution operation is created - however, the *dilated* convolution operation also contains padding dependent on the size of the first input to `call`.
There's no reason for this, as the padding operation (i.e. the call to `tf.nn.conv2d`) could simply be different for each of the two `call`s.
Note that with the functional layer API it is still possible to do the reuse of the layers:
```
conv1 = tf.layers.conv2d(example, 4, 3, padding="SAME", dilation_rate=4, name="test")
conv2 = tf.layers.conv2d(example2, 4, 3, padding="SAME", dilation_rate=4, name="test", reuse=True)
```
because this causes two separate layer creations. I am proposing that the behaviour should be similar to this.

The consequence of this is that at the moment, to my knowledge, there is no good way of doing weight reuse of dilated convolution weights using `tf.keras.layers`. I think that the above is possibly a bug in the implementation - the convolution operation should maybe be create in the `call` function of `tf.keras.layers.Conv`?

**Will this change the current api? How?**
This should not raise an error:
```
import tensorflow as tf

example = tf.zeros([1, 32, 32, 1])
example2 = tf.zeros([1, 41, 41, 1])

convolution_op = tf.layers.Conv2D(4, 3, padding="SAME", dilation_rate=4)

print(convolution_op( example ))
print(convolution_op( example2 ))

```
**Who will benefit with this feature?**
Everyone, as this would be a more consistent behavior of the API, and would allow things that right now are not possible.

**Any Other info.**
More information is also in https://github.com/tensorflow/tensorflow/issues/23019 - I had raised it but it was closed without a relevant reply.

Have I written custom code 
as above.
OS Platform and Distribution
N/A
TensorFlow installed from
pip3
Bazel version
N/A
CUDA/cuDNN version
N/A
GPU model and memory
N/A
Exact command to reproduce
as above
Mobile device
N/A

**Comment by tensorflowbutler** [comment]
Thank you for your post. We noticed you have not filled out the following field in the issue template. Could you update them if they are relevant in your case, or leave them as N/A? Thanks.
Have I written custom code
OS Platform and Distribution
TensorFlow installed from
Bazel version
CUDA/cuDNN version
GPU model and memory
Exact command to reproduce
Mobile device

**Comment by fabio12345** [comment]
Updated.

**Comment by tensorflowbutler** [comment]
Nagging Assignee @fchollet: It has been 15 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by woodshop** [comment]
I also ran into this problem (using TF 1.14.0 and 2.0-compatible operations). I would be curious if this behavior is considered intentional by developers.

For others who land on this issue, I've pasted an example workaround for tf.keras.layers.Conv1D:

```python
import functools
import tensorflow._api.v1.compat.v2 as tf


class Conv1D(tf.keras.layers.Conv1D):
    """Prevent keras from caching the input shape when building.
    """
    # pylint: disable=no-name-in-module
    from tensorflow.python.keras.utils import conv_utils
    # pylint: enable=no-name-in-module

    def build(self, input_shape):
        """Use the functional convolution operation instead of an instance 
        of the internal TF Convolution class, which stores and reuses the input shape.
        """
        super().build(input_shape)

        if self.padding == 'causal':
            op_padding = 'valid'
        else:
            op_padding = self.padding
        if not isinstance(op_padding, (list, tuple)):
            op_padding = op_padding.upper()

        # pylint: disable=attribute-defined-outside-init
        self._convolution_op = functools.partial(
            tf.nn.convolution,
            strides=self.strides,
            padding=op_padding,
            data_format=self.conv_utils.convert_data_format(
                self.data_format, self.rank + 2
            ),
            dilations=self.dilation_rate
        )
        # pylint: enable=attribute-defined-outside-init
```

**Comment by ymodak** [comment]
This is fixed with TF 2.2 
`convolution_op = tf.keras.layers.Conv2D(4, 3, padding="SAME", dilation_rate=4)`
Thanks!
